spec/features/applications/show_spec.rb

```diff
@@ -0,0 +1,177 @@
+      expect(page).to have_button("Search")
+
+      fill_in(:pet_name, with: "Toaster")
+
+      click_button "Search"
+
+      expect(current_path).to eq("/applications/#{@app_1.id}")
+      within "#show-#{@pet_4.id}" do 
+        expect(page).to have_content("Toaster")
+        expect(page).to have_button("Adopt this pet")
+      end
+    end
+
+    it "won't respond to non existent dog name" do
+      visit "/applications/#{@app_1.id}"
+      fill_in(:pet_name, with: "Hosehead")
+      click_button "Search"
+
+      expect(page).to_not have_content("Hosehead")
+    end
+
+    it "has a button to adopt a pet next to each pet search result" do
+      visit "/applications/#{@app_1.id}"
+      fill_in(:pet_name, with: "Hoser")
+
+      click_button "Search"
+      
+      expect(page).to have_button("Adopt this pet")
+      
+      click_button("Adopt this pet")
+      
+      expect(current_path).to eq("/applications/#{@app_1.id}")
+      within '#show-pets-on-app' do
+        expect(page).to have_content("Hoser")
+      end
+    end
+
+    it "has an 'Adopt this Pet' button next each pet" do 
+      
+    end
+
+    it "has a section to submit application after adding pets" do
+      visit "/applications/#{@app_2.id}"
+      expect(page).to_not have_button("Submit this application")
+      expect(page).to_not have_content("Why would you make a good owner?")
+      expect(page).to_not have_field("Howdy")
+
+      expect(@app_2.status).to eq("In progress")
+
+      fill_in(:pet_name, with: "Lobster")
+      click_button "Search"
+      click_button("Adopt this pet")
+
+      fill_in(:pet_name, with: "Hoser")
+      click_button "Search"
+      click_button("Adopt this pet")
+      
+      expect(page).to have_field(:good_owner)
+      expect(page).to have_content("Why would you make a good owner?")
+      fill_in(:good_owner, with: "I'm lonely but also cool")
+
+      expect(page).to have_button("Submit this application")
+      click_button("Submit this application")
+
+      expect(current_path).to eq("/applications/#{@app_2.id}")
+      
+      expect(page).to have_content("Hoser")
+      expect(page).to have_content("Lobster")
+      
+      expect(page).to have_content("Pending")
+      
+      expect(page).to_not have_button("Adopt this pet")
+      expect(page).to_not have_button("Submit this application")
+      expect(page).to_not have_button("Search")
+      expect(page).to_not have_content("Add a Pet to this Application")
+      expect(page).to_not have_field(:pet_name)
+      expect(page).to_not have_content("Search for pet")
+    end
+
+    it "has a partial match when searching for available pets" do
+      visit "/applications/#{@app_2.id}"
+      expect(page).to have_field(:pet_name)
+
+      fill_in(:pet_name, with: "Lob")
+      click_button "Search"
+
+      expect(Pet.search("Lob")).to eq([@pet_2])
+      expect(page).to have_content("Lobster")
+    end
+
+    it "is not case sensitive when searching for available pets" do
```

Comment:
Back in week 1, when we talked about unit test vs integration test, I mentioned the idea of the test pyramid. This is something you'll likely explore later on in the job, but I think it's worth mentioning here. You're doing an awesome job testing this search functionality. There are a bunch of edge cases around non-existent names, partial matches, etc. This is a good example of a case where it can be kind of a pain test test all the edge cases at both the integration and the unit level. It's important to test at both levels, but generally, when following the test pyramid, it's okay to focus on the edge cases more at the unit level and then just verify expected behavior (aka happy path) at the integration level. That means that all the different edge cases for this method could live at the model spec level (i.e. test Pet.search with all these different cases) and then you could just test it with 1-2 happy paths at the integration level. You don't have to have 4-5 edge case tests at *both* levels (unless you really want to!)

Reply:
https://martinfowler.com/articles/practical-test-pyramid.html